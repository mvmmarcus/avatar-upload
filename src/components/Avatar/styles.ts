import styled, { css } from "styled-components";
import media from "styled-media-query";

type AvatarImgProps = {
  scale?: number;
  hasError?: boolean;
  isCropping?: boolean;
};

const avatarImgModifiers = {
  isCropping: (scale?: number) => css`
    transform: ${`scale(${Number(scale)})`};
  `,
};

export const Wrapper = styled.figure`
  ${({ theme }) => css`
    width: 8.4rem;
    height: 8.4rem;
    min-height: 8.4rem;
    min-width: 8.4rem;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 7.25rem;
    overflow: hidden;
    margin: auto;
    position: relative;
    pointer-events: none;
    background-color: ${theme.colors.gray04};

    ${media.greaterThan("small")`
      width: 11.4rem;
      height: 11.4rem;
      min-height: 11.4rem;
      min-width: 11.4rem;
    `}
  `}
`;

export const AvatarImg = styled.img<AvatarImgProps>`
  ${({ scale, isCropping, hasError }) => css`
    width: ${hasError ? "auto" : "100%"};

    transition: transform 0.3s ease-in-out;

    ${isCropping && avatarImgModifiers.isCropping(scale)}
  `}
`;
