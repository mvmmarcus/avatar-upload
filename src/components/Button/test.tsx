import React from "react";

import { screen } from "@testing-library/react";

import Button from ".";
import theme from "styles/theme";
import { renderWithTheme } from "utils/test-utils";

describe("<Button />", () => {
  it("should render default button", () => {
    const { container } = renderWithTheme(<Button>Save</Button>);

    expect(screen.getByRole("button", { name: /Save/i })).toHaveStyle({
      "background-color": theme.colors.gray07,
    });

    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render link button", () => {
    const { container } = renderWithTheme(<Button variant="link">Save</Button>);

    expect(screen.getByRole("button", { name: /Save/i })).toHaveStyle({
      background: "transparent",
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
