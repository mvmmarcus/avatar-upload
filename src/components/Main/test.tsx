import React from "react";

import { screen } from "@testing-library/react";

import Main from ".";
import { renderWithTheme } from "utils/test-utils";

describe("<Main />", () => {
  it("should render main component", () => {
    const { container } = renderWithTheme(<Main>Main</Main>);

    expect(screen.getByText(/main/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
