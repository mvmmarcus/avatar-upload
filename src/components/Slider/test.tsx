import React from "react";

import { fireEvent, screen } from "@testing-library/react";

import Slider from ".";
import { renderWithTheme } from "utils/test-utils";

describe("<Slider />", () => {
  it("should render slider as default", () => {
    const { container } = renderWithTheme(<Slider />);

    expect(screen.getByText(/crop/i)).toBeInTheDocument();
    expect(screen.getByTestId("slider-input")).toHaveProperty("value", "1");
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should change slider value", () => {
    const onValueChange = jest.fn();

    const { container } = renderWithTheme(
      <Slider onValueChange={onValueChange} />
    );

    const sliderInput = screen.getByTestId("slider-input");

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });

    expect(onValueChange).toHaveBeenCalledWith(1.5);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should save image crop", () => {
    const onSave = jest.fn();

    const { container } = renderWithTheme(<Slider onSave={onSave} />);

    const sliderInput = screen.getByTestId("slider-input");
    const saveButton = screen.getByRole("button", { name: /save/i });

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });

    fireEvent.click(saveButton);

    expect(onSave).toHaveBeenCalledWith(1.5);

    expect(container.firstChild).toMatchSnapshot();
  });
});
