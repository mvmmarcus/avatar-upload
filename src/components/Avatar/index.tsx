import React from "react";

import { AvatarImg, Wrapper } from "./styles";

type AvatarProps = {
  isCropping?: boolean;
  urlImg?: string;
  scale?: number;
  hasError?: boolean;
};

const Avatar = ({
  urlImg,
  scale = 1,
  isCropping = false,
  hasError = false,
}: AvatarProps) => {
  return (
    <Wrapper>
      <AvatarImg
        hasError={hasError}
        isCropping={isCropping}
        scale={scale}
        src={urlImg}
        alt="avatar image"
      />
    </Wrapper>
  );
};

export default Avatar;
