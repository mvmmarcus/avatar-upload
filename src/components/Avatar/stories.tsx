import React from "react";

import { Story, Meta } from "@storybook/react/types-6-0";

import Avatar, { AvatarProps } from ".";

export default {
  title: "Avatar",
  component: Avatar,
  argTypes: {
    urlImg: {
      type: "string",
    },
    scale: {
      control: {
        disable: true,
      },
    },
    isCropping: {
      control: {
        disable: true,
      },
    },
  },
} as Meta<AvatarProps>;

export const Default: Story<AvatarProps> = (args) => <Avatar {...args} />;

Default.args = {
  urlImg:
    "https://images.unsplash.com/photo-1494790108377-be9c29b29330?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=387&q=80",
};

export const Error: Story<AvatarProps> = (args) => <Avatar {...args} />;

Error.args = {
  urlImg: "",
};
