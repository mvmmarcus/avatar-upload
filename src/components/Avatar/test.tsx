import React from "react";

import { screen } from "@testing-library/react";

import Avatar from ".";
import { renderWithTheme } from "utils/test/test-utils";

describe("<Avatar />", () => {
  it("should render avatar with specific url", () => {
    const { container } = renderWithTheme(
      <Avatar urlImg="image_url" isCropping={false} />
    );

    expect(screen.getByRole("img", { name: /avatar image/i })).toHaveAttribute(
      "src",
      "image_url"
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render avatar with diferent scale", () => {
    const { container } = renderWithTheme(
      <Avatar urlImg="image_url" scale={1.5} isCropping />
    );

    expect(screen.getByRole("img", { name: /avatar image/i })).toHaveAttribute(
      "scale",
      "1.5"
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render avatar with error", () => {
    const { container } = renderWithTheme(
      <Avatar urlImg="/img/error.svg" hasError />
    );

    expect(screen.getByRole("img", { name: /avatar image/i })).toHaveStyle(
      "width: auto"
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
