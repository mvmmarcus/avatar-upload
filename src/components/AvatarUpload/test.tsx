/* eslint-disable @typescript-eslint/no-explicit-any */
import React from "react";

import { fireEvent, screen, waitFor } from "@testing-library/react";

import AvatarUpload from ".";
import theme from "styles/theme";
import { renderWithTheme } from "utils/test/test-utils";

import * as imageUtils from "utils/image/image";

describe("<AvatarUpload />", () => {
  jest.spyOn(imageUtils, "uploadImage").mockResolvedValue("image_url");
  jest.spyOn(imageUtils, "cropImage").mockResolvedValue("image_url");

  const file = new File(["file"], "image.png", {
    type: "image/png",
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render AvatarUpload as default", () => {
    const { container } = renderWithTheme(<AvatarUpload />);

    expect(container.firstChild).toHaveStyle({ border: theme.border.dashed });
    expect(
      screen.getByRole("img", { name: /media icon/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Organization Logo/i })
    ).toBeInTheDocument();
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();
  });

  it("should invoke dragover and dragleave events", async () => {
    const data = mockDropData([file]);

    const { container } = renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    fireEvent.dragOver(dropZone, data);
    expect(container.firstChild).toHaveStyle({
      border: theme.border.activeDashed,
    });

    fireEvent.dragLeave(dropZone, data);
    expect(container.firstChild).toHaveStyle({
      border: theme.border.dashed,
    });
  });

  it("should drop image", async () => {
    const data = mockDropData([file]);

    renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(screen.getByText("Crop")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();
  });
  it("should get error on upload image", async () => {
    jest.spyOn(imageUtils, "uploadImage").mockImplementation(() => {
      return Promise.reject();
    });

    const data = mockDropData([file]);

    renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(screen.getByText("Sorry, the upload failed.")).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /try again/i })
    ).toBeInTheDocument();
  });
  it("should get an error when trying to drop a file with unsupported extension", async () => {
    const pdfFile = new File(["file"], "file.pdf", {
      type: "file.pdf",
    });
    const data = mockDropData([pdfFile]);

    renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(
      screen.getByText("Sorry, this file extension is not supported.")
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /try again/i })
    ).toBeInTheDocument();
  });
  it("should cancel image cropping", async () => {
    jest.spyOn(imageUtils, "uploadImage").mockResolvedValue("image_url");

    const data = mockDropData([file]);

    renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(screen.getByText("Crop")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();

    fireEvent.click(screen.getByAltText("cancel button"));

    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();
  });

  it("should finish image cropping", async () => {
    const onSave = jest.fn();

    const data = mockDropData([file]);

    renderWithTheme(<AvatarUpload onSave={onSave} />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    fireEvent.change(screen.getByTestId("slider-input"), {
      target: {
        value: 1.5,
      },
    });

    expect(screen.getByRole("img", { name: /avatar image/i })).toHaveAttribute(
      "scale",
      "1.5"
    );

    await waitFor(() => {
      fireEvent.click(screen.getByRole("button", { name: /save/i }));
    });

    expect(onSave).toBeCalledWith("image_url");
    expect(screen.getByRole("img", { name: /avatar image/i })).toHaveAttribute(
      "src",
      "image_url"
    );
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();
  });

  it("should browse image when click on drop zone", async () => {
    const data = mockBrowseData([file]);

    renderWithTheme(<AvatarUpload />);

    const fileInput = screen.getByTestId("file-input");

    fireEvent.click(fileInput);

    await waitFor(() => {
      fireEvent.change(fileInput, data);
    });

    expect(screen.getByText("Crop")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();
  });

  it("should can drop image only on draggable states", async () => {
    const data = mockDropData([file]);

    renderWithTheme(<AvatarUpload />);

    const dropZone = screen.getByTestId("drop-area");

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(screen.getByText("Crop")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();

    await waitFor(() => {
      fireEvent.drop(dropZone, data);
    });

    expect(screen.getByText("Crop")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();
  });
});

function mockDropData(files: File[]) {
  return {
    dataTransfer: {
      files,
      items: files.map((file) => ({
        kind: "file",
        type: file.type,
        getAsFile: () => file,
      })),
      types: [],
    },
  };
}

function mockBrowseData(files: File[]) {
  return {
    target: {
      files,
      items: files.map((file) => ({
        kind: "file",
        type: file.type,
        getAsFile: () => file,
      })),
      types: [],
    },
  };
}
