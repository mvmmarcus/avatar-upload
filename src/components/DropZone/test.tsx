import React from "react";

import { fireEvent, screen } from "@testing-library/react";

import DropZone from ".";
import { renderWithTheme } from "utils/test/test-utils";

describe("<Slider />", () => {
  it("should render DropZone as default", () => {
    const { container } = renderWithTheme(<DropZone />);

    expect(
      screen.getByRole("img", { name: /media icon/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Organization Logo/i })
    ).toBeInTheDocument();
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should change image scale and finish croppping", () => {
    const onFinishCrop = jest.fn();
    const onScaleChange = jest.fn();

    const { container } = renderWithTheme(
      <DropZone
        onFinishCrop={onFinishCrop}
        onScaleChange={onScaleChange}
        isCropping
      />
    );

    const sliderInput = screen.getByTestId("slider-input");
    const saveButton = screen.getByRole("button", { name: /save/i });

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });
    fireEvent.click(saveButton);

    expect(onScaleChange).toHaveBeenCalledWith(1.5);
    expect(onFinishCrop).toHaveBeenCalledWith();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with error and click on try again button", () => {
    const onFinishCrop = jest.fn();

    const { container } = renderWithTheme(
      <DropZone
        onTryAgain={onFinishCrop}
        errorMessage="Sorry, the upload failed."
      />
    );

    expect(screen.getByText("Sorry, the upload failed.")).toBeInTheDocument();
    const tryAgainButton = screen.getByRole("button", { name: /try again/i });

    fireEvent.click(tryAgainButton);

    expect(onFinishCrop).toHaveBeenCalledWith();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with success", () => {
    const { container } = renderWithTheme(<DropZone avatarUrl="image_url" />);

    expect(container.firstChild).toHaveStyle("align-items: center");
    expect(
      screen.getByRole("img", { name: /media icon/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Organization Logo/i })
    ).toBeInTheDocument();
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
