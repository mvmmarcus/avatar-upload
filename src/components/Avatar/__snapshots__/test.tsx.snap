// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> should render avatar with diferent scale 1`] = `
.c0 {
  width: 8.4rem;
  height: 8.4rem;
  min-height: 8.4rem;
  min-width: 8.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 7.25rem;
  overflow: hidden;
  margin: auto;
  position: relative;
  pointer-events: none;
  background-color: #C3CBD5;
}

.c1 {
  width: 100%;
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  -webkit-transform: scale(1.5);
  -ms-transform: scale(1.5);
  transform: scale(1.5);
}

@media (min-width:450px) {
  .c0 {
    width: 11.4rem;
    height: 11.4rem;
    min-height: 11.4rem;
    min-width: 11.4rem;
  }
}

<figure
  class="c0"
>
  <img
    alt="avatar image"
    class="c1"
    scale="1.5"
    src="image_url"
  />
</figure>
`;

exports[`<Avatar /> should render avatar with error 1`] = `
.c0 {
  width: 8.4rem;
  height: 8.4rem;
  min-height: 8.4rem;
  min-width: 8.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 7.25rem;
  overflow: hidden;
  margin: auto;
  position: relative;
  pointer-events: none;
  background-color: #C3CBD5;
}

.c1 {
  width: auto;
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
}

@media (min-width:450px) {
  .c0 {
    width: 11.4rem;
    height: 11.4rem;
    min-height: 11.4rem;
    min-width: 11.4rem;
  }
}

<figure
  class="c0"
>
  <img
    alt="avatar image"
    class="c1"
    scale="1"
    src="/img/error.svg"
  />
</figure>
`;

exports[`<Avatar /> should render avatar with specific url 1`] = `
.c0 {
  width: 8.4rem;
  height: 8.4rem;
  min-height: 8.4rem;
  min-width: 8.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 7.25rem;
  overflow: hidden;
  margin: auto;
  position: relative;
  pointer-events: none;
  background-color: #C3CBD5;
}

.c1 {
  width: 100%;
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
}

@media (min-width:450px) {
  .c0 {
    width: 11.4rem;
    height: 11.4rem;
    min-height: 11.4rem;
    min-width: 11.4rem;
  }
}

<figure
  class="c0"
>
  <img
    alt="avatar image"
    class="c1"
    scale="1"
    src="image_url"
  />
</figure>
`;
